server/routes/message.router.js

```diff
@@ -0,0 +1,124 @@
+const express = require('express');
+const pool = require('../modules/pool');
+const router = express.Router();
+const {
+  rejectUnauthenticated,
+} = require('../modules/authentication-middleware');
+
+router.get('/all', rejectUnauthenticated, (req, res) => {
+  // GET ROUTE - Gets all messages for currently logged-in user
+  const user = req.user.id;
+  const queryText = `
+    SELECT DISTINCT 
+      ON(GREATEST("fromUser", "toUser"), LEAST("fromUser", "toUser")) 
+      GREATEST("fromUser", "toUser"), 
+      LEAST("fromUser", "toUser"), 
+      "date" AS "maxDate", 
+      "message",
+      "users"."firstName",
+      "users"."lastName",
+      "vendors"."companyName",
+      "vendors"."profilePic"
+    FROM "messages"
+    JOIN "users"
+      ON "messages"."fromUser" = "users"."id" OR "messages"."toUser" = "users"."id"
+    RIGHT OUTER JOIN "vendors"
+      On "users"."id" = "vendors"."vendorUserId"
+    WHERE "fromUser" = $1
+      OR "toUser" = $1
+    ORDER BY GREATEST("fromUser", "toUser"), LEAST("fromUser", "toUser"), "maxDate" ASC;
+  `;
+
+  pool
+    .query(queryText, [user])
+    .then((dbRes) => {
+      console.log('SERVER - GET - at /api/message/all - received a response');
+      res.send(dbRes.rows);
+    })
+    .catch((err) => {
+      console.error(
+        'SERVER - GET at /api/message/all - an error occurred',
+        err
+      );
+      res.sendStatus(500);
+    });
+});
+
+router.get('/:id', rejectUnauthenticated, (req, res) => {
+  // GET ROUTE - Gets Messages between two users
+  const fromUser = req.user.id;
+  console.log('fromUser', fromUser);
+  const toUser = req.params.id;
+  console.log('toUser', toUser);
+  const queryText = ` SELECT * FROM "messages" WHERE 
+  "messages"."fromUser" = $1 AND "messages"."toUser" = $2
+  OR "messages"."toUser" = $1 AND "messages"."fromUser" =$2;`;
+
+  pool
+    .query(queryText, [fromUser, toUser])
+    .then((dbRes) => {
+      console.log('SERVER - GET at /api/message/id - received a response');
+      res.send(dbRes.rows);
+    })
+    .catch((err) => {
+      console.error('SERVER - GET at /api/message/id - an error occurred', err);
+      res.sendStatus(500);
+    });
+});
+
+router.post('/', (req, res) => {
+  const queryText = `
+  INSERT INTO "messages" ("fromUser", 
+  "toUser", 
+  "date", 
+  "message")
+  VALUES ($1, $2, $3, $4);`;
+
+  console.log('req.body', req.body);
+
+  pool
+    .query(queryText, [
+      req.body.fromUser,
```

Comment:
**CHANGE REQUEST**
Missing authorization.
The client should not be able to specify the `fromUser` -- would allow any user to send a message, pretending it's from any other user. Ask @eschwartz if you need clarification

Reply:
Completed - John Shands